### ISSUE TITLE
error if a legacy typevar with explicit variance is used in a way that contradicts its explicit variance

### ISSUE THREAD

**Issue by carljm** [astral-sh/ty#1017]
Example:

```py
from typing import TypeVar

T = TypeVar("T", covariant=True)

class C[T]:
    def method(self, x: T):  # this usage requires C to be at least contravariant in T
        pass

**Comment by DetachHead** [comment]
fyi pyrefly and pyright also report an error specifically on `Protocol`s when the explicit variance is *stricter* than its usage, for some reason:

https://pyrefly.org/en/docs/error-kinds/#variance-mismatch

```py
from typing import Protocol, TypeVar

T = TypeVar("T")

class A(Protocol[T]):  # variance-mismatch: Type variable `T` in class `A` is declared as invariant, but could be covariant based on its usage
    def f(self) -> T: ...
```

to me this seems more like a lint rule though, and i don't get why it's only reported on `Protocol`s
